core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -124,20 +124,28 @@ protected Result buildResultAndLogHealthChange(RoutingAllocation allocation, Str
     }
 
     public Result applyFailedShard(ClusterState clusterState, ShardRouting failedShard) {
-        return applyFailedShards(clusterState, Collections.singletonList(new FailedRerouteAllocation.FailedShard(failedShard, null, null)));
+        return applyFailedShards(clusterState, Collections.singletonList(new FailedRerouteAllocation.FailedShard(failedShard, null, null)),
+            Collections.emptyList());
+    }
+
+    public Result applyFailedShards(ClusterState clusterState, List<FailedRerouteAllocation.FailedShard> failedShards) {
+        return applyFailedShards(clusterState, failedShards, Collections.emptyList());
     }
 
     /**
      * Applies the failed shards. Note, only assigned ShardRouting instances that exist in the routing table should be
-     * provided as parameter and no duplicates should be contained.
+     * provided as parameter.
```

Comment:
Can we update the javadocs here to reflect this method also takes care of stale shards by removing their allocation ids?